docs_src/design/adr/security/0008-Secret-Creation-and-Distribution.md

```diff
@@ -0,0 +1,281 @@
+   Secrets distributed in this manner become part of the RaftDB,
+   and thus it becomes necessary to enable swarm autolock mode,
+   which prevents the Raft database encryption key
+   from being stored plaintext on disk.
+
+   * Kubernetes:
+   Kubernetes is not officially supported by the EdgeX project.
+   Kubernetes also supports the secrets volume approach,
+   though the secrets volume can be mounted anywhere in the container namespace.
+   For an example of Kubernetes secrets volumes, see the
+   [Kubernetes secrets documentation](https://kubernetes.io/docs/concepts/configuration/secret/).
+   Secrets distributed in this manner become part of the `etcd` database,
+   and thus it becomes necessary to specify a
+   [KMS provider for data encryption](https://kubernetes.io/docs/tasks/administer-cluster/kms-provider/)
+   to prevent `etcd` from storing plaintext versions of secrets.
+
+   * Docker:
+   Use of a RAM-based temporary file system is desirable to protect secrets,
+   but only Linux platforms support `tmpfs` volumes for secret sharing:
+   generally, `/run` on a Linux system is a `tmpfs` volume.
+   For reasons of portability, `/tmp/edgex/secrets` has been chosen as a
+   cross-platform alternative to `/run`:
+   on the MacOS Docker implementation,
+   `/tmp` is a white-listed shared host directory,
+   while `/run` is not.
+   For production hardening, `/tmp` should be replaced by `/run`
+   where it appears in `docker-compose` files.
+   If the container runtime in use does not support host fileystem bind mounts,
+   then a separate Docker secrets volume should be created for each service.
+
+   * Snaps:
+   Note that due to an existing [bug](https://bugs.launchpad.net/snap-confine/+bug/1620442),
+   `$XDG_RUNTIME_DIR` is set to a snap-specific path
+   (`/run/user/1000/snap.`_$SNAP_NAME_`/`),
+   but the directory is **neither created nor accessible to the snap**.
+   Because it may be necessary to share several secrets outside of the snap
+   --such as secret store tokens--
+   it is desirable to place them in a location that can be mapped to
+   a snap content interface, such as `$SNAP_DATA/secrets`.
+   Although not a `tmpfs`, this is the best available option at this time.
+   The snap bug should be escalated in priority.
+
+## Consequences
+
+As the existing implementation is not fully-compliant with this ADR,
+significant scope will be added to current and future EdgeX releases
+in order to bring the project into compliance.
+
+List of needed improvements:
+
+- PKI private keys
+  * All: Move to using Vault as system of origin for the PKI instead of the standalone `security-secrets-setup` utility.
+  * All: Cache the PKI for Consul and Vault on persistent disk; rotate occasionally.
+  * All: Investigate hardware protection of cached Consul and Vault PKI secret keys.
+  * Snaps: Relocate PKI into the secrets volume instead of its own folder.
```

Comment:
There's no concept of volumes in snaps. The PKI materials are all written to directories under `$SNAP_DATA/currrent/secrets`.

Reply:
> There's no concept of volumes in snaps. The PKI materials are all written to directories under `$SNAP_DATA/currrent/secrets`.

I would like a suggestion for a name to use in place of "secrets volume."

If supported, the ideal place for the **to-be-named location** would be `/run/user/1000/snap._$SNAP_NAME_/secrets`, falling back to `_$SNAP_DATA_/currrent/secrets`.

Reply:
> There's no concept of volumes in snaps. The PKI materials are all written to directories under `$SNAP_DATA/currrent/secrets`.

@tonyespy if you could come up with a short name that is "the place that secrets are stored in a snap" that would be great.

Reply:
I've called this _SECRETSLOC_ in the document for lack of a better name.

Reply:
I just confirmed that **ca.pem**, and the **server.crt** and **server.key** files are being re-generated every "cold restart of the framework" in the snap, so with default configuration it might make sense to change the SECRETSLOC to `/run/snap.$SNAP_NAME/secrets` (note the lack of `/user/id` in the path as this doesn't really make sense in the context of a snap).

That said, I'd really like to see us address the issue of external certificates as part of Hanoi, as our current self-generated root CA approach doesn't really work well in production environments (https://github.com/edgexfoundry/edgex-go/issues/1926). If this were to be implemented, it wouldn't make sense to use a tmpfs based directory...

I also noticed that the **server.crt** and **server.key** files are the same for kong and vault within the snap, so this might be a bug in the configuration.

Reply:
> I just confirmed that **ca.pem**, and the **server.crt** and **server.key** files are being re-generated every "cold restart of the framework" in the snap, so with default configuration it might make sense to change the SECRETSLOC to `/run/snap.$SNAP_NAME/secrets` (note the lack of `/user/id` in the path as this doesn't really make sense in the context of a snap).

I have done `snap run --env hello-world` followed by `env` command.  Filtering for `/run` I get:

```
GPG_AGENT_INFO=/run/user/1000/gnupg/S.gpg-agent:0:1
SSH_AUTH_SOCK=/run/user/1000/keyring/ssh
PWD=/run
DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/1000/bus
XDG_RUNTIME_DIR=/run/user/1000/snap.hello-world
XAUTHORITY=/run/user/1000/gdm/Xauthority
```

(BTW: XDG_RUNTIME_DIR doesn't exist either)

What variable is supposed to map to `/run/snap.hello-world`

Reply:
> That said, I'd really like to see us address the issue of external certificates as part of Hanoi, as our current self-generated root CA approach doesn't really work well in production environments ([edgexfoundry/edgex-go#1926](https://github.com/edgexfoundry/edgex-go/issues/1926)). If this were to be implemented, it wouldn't make sense to use a tmpfs based directory...

I think this is pretty simple to fix.  I added language to the document.  Summary: it's already in the secret store.  Just add a flag to keep the framework from overwriting it and a utility to install one manually.

> I also noticed that the **server.crt** and **server.key** files are the same for kong and vault within the snap, so this might be a bug in the configuration.

Hmmm.  Let me know.